GPU/Software/Lighting.cpp

```diff
@@ -86,70 +87,83 @@ void Process(VertexData& vertex, bool hasColor) {
-			float rawSpot = Dot(dir.NormalizedOr001(cpu_info.bSSE4_1), L);
+			float rawSpot = Dot(dir.Normalized(cpu_info.bSSE4_1), L);
+			if (std::isnan(rawSpot))
+				rawSpot = std::signbit(rawSpot) ? 0.0f : 1.0f;
 			float cutoff = getFloat24(gstate.lcutoff[light]);
+			if (std::isnan(cutoff) && std::signbit(cutoff))
+				cutoff = 0.0f;
 			if (rawSpot >= cutoff) {
 				float conv = getFloat24(gstate.lconv[light]);
 				spot = pspLightPow(rawSpot, conv);
+				if (std::isnan(spot))
+					spot = 0.0f;
 			} else {
-				spot = 0.f;
+				spot = 0.0f;
 			}
 		}
 
 		// ambient lighting
-		Vec3<float> lac = Vec3<float>::FromRGB(gstate.getLightAmbientColor(light));
-		final_color += lac * mac * att * spot;
+		int attspot = (int)ceilf(256 * 2 * att * spot + 1);
+		if (attspot > 512)
+			attspot = 512;
+		Vec4<int> lac = Vec4<int>::FromRGBA(gstate.getLightAmbientColor(light));
+		Vec4<int> lambient = ((mac * 2 + ones) * (lac * 2 + ones) * attspot) / (1024 * 512);
+		final_color += lambient;
 
 		// diffuse lighting
-		Vec3<float> ldc = Vec3<float>::FromRGB(gstate.getDiffuseColor(light));
-		Vec3<float> mdc = (materialupdate & 2) ? vcol0 : Vec3<float>::FromRGB(gstate.getMaterialDiffuse());
-
 		float diffuse_factor = Dot(L, vertex.worldnormal);
 		if (gstate.isUsingPoweredDiffuseLight(light)) {
 			float k = gstate.getMaterialSpecularCoef();
 			diffuse_factor = pspLightPow(diffuse_factor, k);
 		}
 
-		if (diffuse_factor > 0.f) {
-			final_color += ldc * mdc * diffuse_factor * att * spot;
+		if (diffuse_factor > 0.0f) {
+			int diffuse_attspot = (int)ceilf(256 * 2 * att * spot * diffuse_factor + 1);
+			if (diffuse_attspot > 512)
+				diffuse_attspot = 512;
+			Vec4<int> ldc = Vec4<int>::FromRGBA(gstate.getDiffuseColor(light));
+			Vec4<int> mdc = (materialupdate & 2) ? vertex.color0 : Vec4<int>::FromRGBA(gstate.getMaterialDiffuse());
+			Vec4<int> ldiffuse = ((ldc * 2 + ones) * (mdc * 2 + ones) * diffuse_attspot) / (1024 * 512);
+			final_color += ldiffuse;
 		}
 
 		if (gstate.isUsingSpecularLight(light) && diffuse_factor >= 0.0f) {
 			Vec3<float> H = L + Vec3<float>(0.f, 0.f, 1.f);
 
-			Vec3<float> lsc = Vec3<float>::FromRGB(gstate.getSpecularColor(light));
-			Vec3<float> msc = (materialupdate & 4) ? vcol0 : Vec3<float>::FromRGB(gstate.getMaterialSpecular());
-
 			float specular_factor = Dot(H.NormalizedOr001(cpu_info.bSSE4_1), vertex.worldnormal);
 			float k = gstate.getMaterialSpecularCoef();
 			specular_factor = pspLightPow(specular_factor, k);
 
-			if (specular_factor > 0.f) {
-				specular_color += lsc * msc * specular_factor * att * spot;
+			if (specular_factor > 0.0f) {
+				int specular_attspot = (int)ceilf(256 * 2 * att * spot * specular_factor + 1);
+				if (specular_attspot > 512)
+					specular_attspot = 512;
+				Vec4<int> lsc = Vec4<int>::FromRGBA(gstate.getSpecularColor(light));
+				Vec4<int> msc = (materialupdate & 4) ? vertex.color0 : Vec4<int>::FromRGBA(gstate.getMaterialSpecular());
```

Comment:
I guess these could be computed and cached outside the loop, although I doubt it'll be a huge win, especially considering how rare specular is.. So probably not heh.

Reply:
I tried this and it didn't seem to help performance, though it was hard to tell.  Mostly seemed to stay the same...

-[Unknown]